frame/bonded-finance/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+					<T::Vesting as VestedTransfer>::MinVestedTransfer::get(),
+					T::MinReward::get()
+				),
+				Error::<T>::InvalidBondOffer
+			);
+			let offer_id = BondOfferCount::<T>::mutate(|offer_id| {
+				*offer_id = offer_id.next();
+				*offer_id
+			});
+			let offer_account = Self::account_id(offer_id);
+			T::NativeCurrency::transfer(from, &offer_account, T::Stake::get(), true)?;
+			T::Currency::transfer(
+				offer.reward_asset,
+				from,
+				&offer_account,
+				offer.reward_amount,
+				true,
+			)?;
+			BondOffers::<T>::insert(offer_id, (from.clone(), offer));
+			Self::deposit_event(Event::<T>::NewOffer { offer: offer_id });
+			Ok(offer_id)
+		}
+
+		#[transactional]
+		pub fn do_bond(
+			offer_id: T::BondOfferId,
+			from: &AccountIdOf<T>,
+			contracts: BalanceOf<T>,
+		) -> Result<BalanceOf<T>, DispatchError> {
+			BondOffers::<T>::try_mutate(offer_id, |offer| {
+				offer
+					.as_mut()
+					.map(|(creator, offer)| {
+						ensure!(
+							offer.contracts > BalanceOf::<T>::zero(),
+							Error::<T>::OfferCompleted
+						);
+						ensure!(
+							contracts > BalanceOf::<T>::zero() && contracts <= offer.contracts,
+							Error::<T>::InvalidNumberOfContracts
+						);
+						// NOTE(hussein-aitlahcen): can't overflow, subsumed by `offer.valid()` in
+						// `do_offer`
+						let value = contracts * offer.price;
+						ensure!(
+							T::Currency::can_withdraw(offer.asset, from, value)
+								.into_result()
+								.is_ok(),
+							Error::<T>::NotEnoughAsset
+						);
+						let offer_account = Self::account_id(offer_id);
+						T::Currency::transfer(offer.asset, from, &offer_account, value, true)?;
+						let reward_share = T::Convert::convert(
+							multiply_by_rational(
+								T::Convert::convert(value),
+								T::Convert::convert(offer.reward_amount),
+								// NOTE(hussein-aitlahcen): checked by `offer.valid()` in
+								// `do_offer`
+								T::Convert::convert(offer.total_price().expect("impossible; qed;")),
+							)
+							.map_err(|_| ArithmeticError::Overflow)?,
+						);
+						let block = frame_system::Pallet::<T>::current_block_number();
+						T::Vesting::vested_transfer(
+							offer.reward_asset,
+							&offer_account,
+							from,
+							VestingSchedule {
+								start: block,
+								period: offer.reward_duration,
+								period_count: 1,
+								per_period: reward_share,
+							},
+						)?;
+						match offer.duration {
+							BondDuration::Finite { blocks } => {
+								T::Vesting::vested_transfer(
+									offer.asset,
+									&offer_account,
+									from,
+									VestingSchedule {
+										start: block,
+										period: blocks,
+										period_count: 1,
+										per_period: value,
+									},
+								)?;
+							},
+							BondDuration::Infinite => {
+								// the liquidity is now owned by us
+							},
+						}
+						(*offer).contracts -= contracts;
+						(*offer).reward_amount -= reward_share;
+						Self::deposit_event(Event::<T>::NewBond {
```

Comment:
may be deposit as last step, losing event is not so bad during write

Reply:
done!